### PR TITLE
Added `box-sizing: border-box;` to...

### PR DESCRIPTION
**Pull request by joshdarby**
`figcaption` and `.wp-caption-text` elements so that the caption text doesn't overflow from the parent container. Issue #1702

### CODE REVIEW

**Comment by joshdarby** [comment]
@benlk Can you review this and let me know if you think we need to add the `box-sizing` attribute to any other versions of `figcaption` and `.wp-caption-text`?

**Comment by benlk** [comment]
`.wp-block-gallery`, which I'll add some styles for in here.

File: less/inc/images.less [hunk 1]
```diff
@@ -28,6 +28,9 @@
         -o-object-fit: scale-down;
         object-fit: scale-down;
       }
+      figcaption {
```
Comment (benlk) [inline less/inc/images.less]: This selector is `.wp-block-gallery.is-cropped .blocks-gallery-item figcaption` but not all galleries are `.is-cropped`.

**Comment by benlk** [comment]
The padding/max-height changes in https://github.com/INN/largo/pull/1711/commits/df15461147fb02201cdacef5afd607098b593da9 are to make sure that with Very Long captions, the caption doesn't fully cover the image.

Screen Shot 2019-06-17 at 19 14 05
Screen Shot 2019-06-17 at 19 14 24

But the Gutenberg gallery styles here don't match anything in Largo: we generally don't overlay captions over images. Should we try to put the captions beneath the image, like happens with the classic editor's gallery shortcode?

Screen Shot 2019-06-17 at 20 00 50


If the gallery shortcode and the classic editor's gallery and the classic editor's image show captions in the expanded slideshow view, should the classic block?

Classic image:
Screen Shot 2019-06-17 at 19 59 11

Gallery block:
Screen Shot 2019-06-17 at 19 59 16

**Comment by joshdarby** [comment]
> But the Gutenberg gallery styles here don't match anything in Largo: we generally don't overlay captions over images. Should we try to put the captions beneath the image, like happens with the classic editor's gallery shortcode?

@benlk I'm not opposed to either way. I'd think if having them below matches Largo styles more, we should go that route.

> If the gallery shortcode and the classic editor's gallery and the classic editor's image show captions in the expanded slideshow view, should the classic block?

If 3/4 of them are showing the caption in the expanded view, I'd think the 4th should show it also. Unless you think there's a use case where someone using the theme would want one instance to use that doesn't show the captions in that view?

**Comment by benlk** [comment]
> > If the gallery shortcode and the classic editor's gallery and the classic editor's image show captions in the expanded slideshow view, should the classic block?
> 
> If 3/4 of them are showing the caption in the expanded view, I'd think the 4th should show it also. Unless you think there's a use case where someone using the theme would want one instance to use that doesn't show the captions in that view?

I don't believe there is such a case; let's make those cases consistent. 

>> But the Gutenberg gallery styles here don't match anything in Largo: we generally don't overlay captions over images. Should we try to put the captions beneath the image, like happens with the classic editor's gallery shortcode?
> 
> @benlk I'm not opposed to either way. I'd think if having them below matches Largo styles more, we should go that route.

My main worry here is that, since we have Gutenberg's default stylesheet enqueued, Gutenberg will change something upstream will horribly conflict with the styles we apply here, and as a result our sites will complain.

I'd love for there to be a way to get Gutenberg to enqueue its styles per-block, so that we would be able to dequeue the Gallery Block styles and replace those, but then let Gutenberg enqueue the styles for New Shiny Block until we can write custom styles for it.

Should we go ahead and make drastic changes to CSS without dequeueing Gutenberg's styles?

**Comment by benlk** [comment]
> Should we go ahead and make drastic changes to CSS without dequeueing Gutenberg's styles?

No.

Therefore, let's not change the gallery block caption display location, yet.

**Comment by benlk** [comment]
Next steps:

- [x] Make sure all captions display consistently in slideshow view.

**Comment by joshdarby** [comment]
> * [ ]  Make sure all captions display consistently in slideshow view.

@benlk Should they all display the way they do in the classic image or in the Gutenberg block?

**Comment by benlk** [comment]
My personal preference is the classic image view, because that's easier to read than the gradient background.

**Comment by joshdarby** [comment]
@benlk Now the captions throughout classic image, image gallery block, and classic gallery shortcode are consistent. The only inconsistency is the known issue of Gutenberg image blocks/galleries not displaying the media credits from issue #1683.

**Comment by benlk** [comment]
Sounds good; can you shove screenshots in this ticket and update the changelog before we merge?

**Comment by joshdarby** [comment]
@benlk 

Classic image block:
Screen Shot 2019-06-24 at 1 42 30 PM

Classic image gallery:
Screen Shot 2019-06-24 at 1 42 13 PM

Gutenberg gallery block:
Screen Shot 2019-06-24 at 1 42 23 PM

File: less/inc/images.less [hunk 2]
```diff
@@ -30,4 +30,19 @@
       }
     }
   }
+  .blocks-gallery-item {
+    .navis-single {
+      figcaption {
+        background-color: transparent;
+        color: #ccc;
+        padding: 1em 17.023%!important;
```
[the remark this answers is not in the document]
Comment (benlk) [reply less/inc/images.less]: It matches https://github.com/INN/largo/blob/c41e28e6381397be1918184b906ec84bc019745b/lib/navis-slideshows/css/slides.css#L310-L312 , but yes, this should be documented in code.

**Comment by benlk** [comment]
The background shouldn't be flat transparent; it should be `rgba(0,0,0,.7);` to match the Navis slideshow styles:

https://github.com/INN/largo/blob/c41e28e6381397be1918184b906ec84bc019745b/lib/navis-slideshows/css/slides.css#L308

Also, the figcaption in wp-block-image in the Navis slideshow state doesn't match these styles; the caption is hidden.

-----

**Edit**: this is fixed now.

**Comment by benlk** [comment]
When testing images, we need to test all of these separate sets of conditions:

- caption state
    1. no caption
    2. caption
    3. media credit
        - name
        - name and org
        - name and org and link
    4. caption and media credit
- image type
    1. classic editor image
    2. classic editor gallery shortcode
    3. block editor image
    4. block editor gallery
- alignment options (block and classic)
    1. unset
    2. left
    3. center
    4. right
    5. wide
    6. full

That's ... 6 * 4 * 6 = 144 different potential image types to test. 288 if we include the clicked-on Navis Slideshow as another variable.

We need a plan for testing this stuff systematically, or at least a stock set of markup to test against that generates all those options.

**Comment by benlk** [comment]
Also, working on this changeset and similar ones has made me doubt the usefulness of splitting the Navis Slideshow CSS out from other image-related CSS in Largo; there's a lot of common variables. I'd be in favor of generating it using the Grunt workflow from LESS, where we could set shared variables between the other image code and the slideshow code. The slideshow CSS/JS should still only be enqueued when needed.

**Comment by benlk** [comment]
Another image case to consider:

- WordPress gallery widget
- Largo image widget
- WordPress image widget

**Comment by joshdarby** [comment]
@benlk Do you want to get together and figure out a plan for acceptable criteria for getting this PR merged?

**Comment by benlk** [comment]
Yeah, let's set aside time tomorrow to test this.

**Comment by benlk** [comment]
IIRC, the plan we had discussed on that day was:

- test against the test content we have available to us now
- postpone the Grand Generation of Test Content until a later release

**Comment by joshdarby** [comment]
@benlk I created a sample post [here](http://largodev.innsandbox.wpengine.com/2019/07/17/box-sizing-examples/) with a 

- [ ] classic block image 
- [ ] classic block gallery
- [ ] image block
- [ ] gallery block

What else do you think we should test for this?

**Comment by benlk** [comment]
Do we know what's going on in Firefox where opening the modal scrolls the browser to the top of the page, triggering the appearance of the sticky nav? https://github.com/INN/largo/issues/1745

Screen Shot 2019-07-17 at 16 14 10

**Comment by benlk** [comment]
There's a difference in how the image block and Gallery block display their images, with whether the images are stretched to full size:

```html
<figure class="wp-block-image navis-slideshow navis-single navis-full" style="max-width: 100%;">
	<span class="navis-before">X</span>
	<img
		src="http://largodev.innsandbox.wpengine.com/wp-content/uploads/2019/07/Screen-Shot-2019-07-17-at-2.48.12-PM.png"
		alt=""
		class="wp-image-1884"
		srcset="http://largodev.innsandbox.wpengine.com/wp-content/uploads/2019/07/Screen-Shot-2019-07-17-at-2.48.12-PM.png 420w, http://largodev.innsandbox.wpengine.com/wp-content/uploads/2019/07/Screen-Shot-2019-07-17-at-2.48.12-PM-140x140.png 140w, http://largodev.innsandbox.wpengine.com/wp-content/uploads/2019/07/Screen-Shot-2019-07-17-at-2.48.12-PM-336x333.png 336w, http://largodev.innsandbox.wpengine.com/wp-content/uploads/2019/07/Screen-Shot-2019-07-17-at-2.48.12-PM-60x60.png 60w, http://largodev.innsandbox.wpengine.com/wp-content/uploads/2019/07/Screen-Shot-2019-07-17-at-2.48.12-PM-96x96.png 96w"
	>
	<figcaption>Largo is a responsive WordPress theme framework designed for news publishers. Finely-crafted and expertly-honed by the product and technology team at the Institute for Nonprofit News.</figcaption>
</figure>

```

Screen Shot 2019-07-17 at 16 16 52


```html
<ul class="wp-block-gallery columns-1 is-cropped">
	<li class="blocks-gallery-item">
		<figure class="navis-slideshow navis-single navis-full" style="max-width: 100%;">
			<span class="navis-before">X</span>
			<img
				src="http://largodev.innsandbox.wpengine.com/wp-content/uploads/2019/07/Screen-Shot-2019-07-17-at-2.48.12-PM.png"
				alt=""
				data-id="1884"
				data-link="http://largodev.innsandbox.wpengine.com/?attachment_id=1884"
				class="wp-image-1884"
				srcset="http://largodev.innsandbox.wpengine.com/wp-content/uploads/2019/07/Screen-Shot-2019-07-17-at-2.48.12-PM.png 420w, http://largodev.innsandbox.wpengine.com/wp-content/uploads/2019/07/Screen-Shot-2019-07-17-at-2.48.12-PM-140x140.png 140w, http://largodev.innsandbox.wpengine.com/wp-content/uploads/2019/07/Screen-Shot-2019-07-17-at-2.48.12-PM-336x333.png 336w, http://largodev.innsandbox.wpengine.com/wp-content/uploads/2019/07/Screen-Shot-2019-07-17-at-2.48.12-PM-60x60.png 60w, http://largodev.innsandbox.wpengine.com/wp-content/uploads/2019/07/Screen-Shot-2019-07-17-at-2.48.12-PM-96x96.png 96w"
			>
			<figcaption>Largo is a responsive WordPress theme framework designed for news publishers. Finely-crafted and expertly-honed by the product and technology team at the Institute for Nonprofit News.</figcaption>
		</figure>
	</li>
</ul>
```
Screen Shot 2019-07-17 at 16 16 57

**Comment by joshdarby** [comment]
> There's a difference in how the image block and Gallery block display their images, with whether the images are stretched to full size:

@benlk I think that's a known issue, right? And is it in the scope of #1702 to fix in this PR?

**Comment by benlk** [comment]
I'm not sure it's a known issue; I'll create a ticket for it and let's merge this. 

https://github.com/INN/largo/issues/1745 isn't in this milestone either.

**Pull request merged**